docs/reference/style-guides/node-style-guide.md

```diff
@@ -2,8 +2,311 @@
+
+```
+app/
+├── auth/
+│   ├── dto/
+│   ├── model/
+│   |   └── auth.model.ts
+│   ├── auth.controller.ts
+│   └── auth.validator.ts
+```
+
+Tests should be organized alongside the code being tested, with the `.spec.ts` suffix or `.in.spec.ts` for integration tests:
+
+```
+app/
+├── auth/
+│   ├── auth.controller.ts
+│   ├── auth.controller.spec.ts
+│   ├── auth.controller.in.spec.ts
+```
+
+#### NestJS File Naming Conventions
+
+##### File Naming for Class
+
+```typescript
+export class PascalCaseSuffix {} //= pascal-case.suffix.ts
+// Except for suffix, PascalCase to hyphen-case
+class FooBarNaming {} //= foo-bar.naming.ts
+class FooController {} //= foo.controller.ts
+// These are both DTO's, which the package namespace should reflect rather than
+// adding DTO to every class name and file name.
+class BarQueryInput {} //= dto/bar-query.input.ts
+class BarResultPayload {} //= dto/bar-result.payload.ts
+
+abstract class BaseAccount {} //= baseAccount.ts
+```
+
+#### Interface and Abstract Naming
+
+- Avoid prefixing with `I` or `Interface`, as this impacts readability and the type of object is separate from the name.
+- Don't include `Abstract` in the name of abstract classes, as this is the type. Prefer `Base` or `Default` prefix.
+
+```typescript
+interface User {}
+interface CustomUser extends User {}
+interface ThirdCustomUser extends CustomUser {}
+abstract class BaseAccount {}
+```
+
+#### Index Exporting
+
+1. It is recommended to place index.ts in each folder and export. Unless it's a special case, it is imported from a folder instead of directly from a file.
+
+Don't:
+
+```typescript
+import { FooController } from './feature/foo.controller';
+import { BarController } from './feature/bar.controller';
+```
+
+Do:
+
+```typescript
+import { FooController, BarController } from './feature';
+```
+
+2. Preferred method is to place only one file or folder name at the end of the path.
+
+Don't:
+
+```typescript
+import { UtilService } from '../common/providers/util.service';
+```
+
+Do:
+
+```typescript
+import { UtilService } from '../common';
+```
+
+### TypeScript
+
+#### Tests
+
+All tests should be written in TypeScript. Tests should utilize factory functions to generate objects for testing. Generating additional fake data for factory functions can be done with the `faker` library.
+
+Example:
+
+```typescript
+import { faker } from '@faker-js/faker';
+
+interface User {
+  id: string;
+  name: string;
+  email: string;
+}
+
+const UserFactory = (override: Partial<User>): User => ({
+  id: faker.datatype.uuid(),
+  name: faker.person.fullName(),
+  email: faker.internet.email(),
+  ...override,
+});
+```
+
+#### Never use Type Asserts
+
+Type asserts should not be used in any situation. Instead, use one of the following options:
+
+1. Check if the variable is null/undefined, and if so throw an error
+
+Do:
+
+```typescript
+function example(arg?: string) {
+  if (!arg) {
```

Comment:
The intent here is clear enough,  but the example is poor. If you pass in an empty string, which is in fact still a string and possibly a valid case, you'll get an exception. I think this hits on an important point to guide developers on. When type checking, it's good practice to be strict. Overly loose checks, will lead to loose type validation, and can result in bugs. 

Another minor nit... the example feels contrived. I don't think you'd see this in the real world, where you allow a type only to generate a runtime error every time that value is passed in. Most the type guard examples I've seen either use one of two scenarios. They show a switch in function calls based on type, or the examples use an `unknown` type, and fail fast like this one. I think either of those scenarios are more realistic.

We should also encourage creating [reusable type guards](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) as part of the guide. Again, these guards need to be robust and tested to avoid problems, but can be very handy when coercing objects that come from external sources.

Reply:
We can also use `class-validator` for handling external objects which can do type validation and additional constraint checks.